### PR TITLE
SMP: update stats and useInView

### PR DESCRIPTION
**Pull request by sejas**
- Fixes https://github.com/Automattic/dotcom-forge/issues/1097

#### Proposed Changes

The issue came from not using a `useCallback`, so the useInView `useEffect` was retriggered in every render, which disconnected the observer, and some callback Stats were not fired when the component was visible.

* Replace `react-intersection-observer` with our custom hook.
* Fix race condition on Stats and improve `useInView`  to fix and show Stats even on fast scrolling.

#### Testing Instructions

* Go to /sites
* Select the row mode (not grid)
* Click on refresh browser with the shift key pressed.
* Scroll fast the sites
* Observe the Stats are visible on every row

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ x Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

### CODE REVIEW

File: client/lib/use-in-view/index.ts [hunk 1]
```diff
@@ -39,7 +39,15 @@ export function useInView< T extends Element >( oneTimeCallback: () => void ): R
...
-		return () => observerRef.current?.disconnect?.();
+		return () => {
+			observerRef.current?.disconnect?.();
+			if ( ! viewedRef.current && observerRef.current ) {
+				// This case prevents when the effect is dismounted and re-mounted due oneTimeCallback is not wrapped in useCallback
+				// We cannot show a warning, because it can be a valid use case.
+				// Valid case: When a component is unmounted and never was visible.
+				observerRef.current = undefined;
+			}
+		};
```
Comment (danielbachhuber) [inline client/lib/use-in-view/index.ts]: Should this have tests?

Comment (sejas) [reply client/lib/use-in-view/index.ts]: Sure, I'll write those tests 👌 .

Comment (danielbachhuber) [reply client/lib/use-in-view/index.ts]: Just so it's stated: I still think we should [use the well-tested dependency](https://github.com/Automattic/wp-calypso/pull/69119#pullrequestreview-1143893128) so we don't have to discover all of these edge cases ourselves.

**Review by vykes-mac — COMMENTED** [review]
The good old `useEffect` race condition. This fix seems reasonable but I still wasn't able to repro on my end.

File: client/lib/use-in-view/index.ts [hunk 2]
```diff
@@ -28,8 +28,8 @@ export function useInView< T extends Element >( oneTimeCallback: () => void ): R
 			}
 
 			oneTimeCallback();
-
 			viewedRef.current = true;
+			observerRef.current?.disconnect?.();
 		};
 
 		observerRef.current = new IntersectionObserver( handler, {
```
Comment (zaguiini) [inline client/lib/use-in-view/index.ts]: Maybe I missed something when implementing this, but what's the point of saving the observer to a ref? Theoretically, if `oneTimeCallback`'s reference is stable, this won't be re-created on every render since the effect will run only once 🤔

File: client/sites-dashboard/components/sites-table-row.tsx [hunk 3]
```diff
@@ -99,7 +99,8 @@ export default memo( function SitesTableRow( { site }: SiteTableRowProps ) {
 	const { __ } = useI18n();
 	const translatedStatus = useSiteLaunchStatusLabel( site );
 	const [ inViewOnce, setInViewOnce ] = useState( false );
-	const ref = useInView< HTMLTableDataCellElement >( () => setInViewOnce( true ) );
+	const setInViewOnceCallback = useCallback( () => setInViewOnce( true ), [] );
```
Comment (zaguiini) [inline client/sites-dashboard/components/sites-table-row.tsx]: We don't need to wrap this into a useCallback hook if we store the callback for `useInView` inside a ref. We can always update the ref on render by calling `ref.current = callback`, so we don't need to introduce more mental models to the user.

Example in `useInView` hook:

```
const callback = useRef();

callback.current = callbackFromHookParameter;
```

Then inside the observer's callback:

```
callback.current?.();
```

This is what the dependency is doing: https://github.com/thebuilder/react-intersection-observer/blob/master/src/useInView.tsx.

Comment (p-jackson) [reply client/sites-dashboard/components/sites-table-row.tsx]: > We don't need to wrap this into a useCallback hook if we store the callback for `useInView` inside a ref

I agree, it's better not to put a requirement on the user of `useInView` hook to have a persistent function reference. It's too easy to forget, and the calling code shouldn't have to think of these sort of implementation details.

File: client/lib/use-in-view/index.ts (hunk 2 above)
Comment (p-jackson) [reply client/lib/use-in-view/index.ts]: Indeed! Just declaring the `IntersectionObserver` directly inside the `useEffect` callback would also remove the need for the cleanup code that's been added to the dismount function.

File: client/sites-dashboard/components/sites-table-row.tsx (hunk 3 above)
Comment (mpkelly) [reply client/sites-dashboard/components/sites-table-row.tsx]: Another +1 for this approach.

**Comment by mpkelly** [comment]
Like @vykes-mac, I also can't reproduce this on prod, no matter how fast I scroll. I can see the bug in the code though. 

As far as I can see, removing this would be enough to fix the problem:

```js
// If the observer is already defined, no need to continue
if ( observerRef.current ) {
	return;
}
```
It doesn't matter if `onTimeCallback` is stable or not because it's in the dependency array and the proper clean-up code is in place. With the snippet above removed we will add/remove the listener properly each time `oneTimeCallback` changes, and not hold on to the stale  `IntersectionObserver` ref from the first time the code ran. 

As @p-jackson noted, we can remove `const observerRef = useRef< IntersectionObserver >();` and just create a locally-scoped instance of `IntersectionObserver` in the `useEffect` function that we can also access in the clean-up function.

File: client/sites-dashboard/components/sites-table-row.tsx (hunk 3 above)
Comment (mpkelly) [reply client/sites-dashboard/components/sites-table-row.tsx]: >Another +1 for this approach.

Actually, on further inspection, I think the fix is even simpler. I added a [comment](https://github.com/Automattic/wp-calypso/pull/69378#issuecomment-1289973670).

File: client/lib/use-in-view/index.ts (hunk 1 above)
Comment (sejas) [reply client/lib/use-in-view/index.ts]: I like our implementation. The `IntersectionObserver` is not something that is evolving constantly, so we don't need to worry about maintenance or finding more edge cases.

Comment (sejas) [reply client/lib/use-in-view/index.ts]: @danielbachhuber, test added here https://github.com/Automattic/wp-calypso/pull/69378/commits/3f13254be8c58146174569be9d21c90f3033f276

**Comment by sejas** [comment]
@Automattic/build I'm really happy to have so many eyes and good opinions. I feel like standing on the shoulders of giants.

I took this approach https://github.com/Automattic/wp-calypso/pull/69378#discussion_r1003603440, so the person who uses this hook shouldn't be worried about wrapping the call on a `useCallback`.

I've also added a test for the issue.

And I'm not able to reproduce the original issue 👌

File: client/lib/use-in-view/test/index.tsx [hunk 4]
```diff
@@ -66,4 +66,14 @@ describe( 'useInView suite', () => {
...
+
+	it( 'Component renders twice, first out view, second in view, event fires once', () => {
+		let count = 0;
+		// The callback is generated in each render, so it's a new function
+		const { rerender } = render(
+			<TestComponent totalTimesInView={ 0 } callback={ () => count++ } />
+		);
+		rerender( <TestComponent totalTimesInView={ 1 } callback={ () => count++ } /> );
+		expect( count ).toBe( 1 );
+	} );
```
Comment (sejas) [inline client/lib/use-in-view/test/index.tsx:85]: Test case added. The component is rendered twice. The first render is not visible, and the second one becomes visible.

File: client/lib/use-in-view/index.ts [hunk 5]
```diff
@@ -1,21 +1,20 @@
...
 export function useInView< T extends Element >( oneTimeCallback: () => void ): RefObject< T > {
-	const viewedRef = useRef< boolean >( false );
+	const viewedRef = useRef( false );
 	const elementRef = useRef< T >( null );
-	const observerRef = useRef< IntersectionObserver >();
+	const oneTimeCallbackRef = useRef< () => void >();
+
+	useEffect( () => {
+		oneTimeCallbackRef.current = oneTimeCallback;
+	}, [ oneTimeCallback ] );
```
Comment (zaguiini) [inline client/lib/use-in-view/index.ts]: This is OK but useless, we can just set the reference on render without the effect since there's not really a performance penalty with this assignment. Furthermore, it might actually prevent race conditions between updated versions of the function and the actual listener inside the effect below. I'd recommend unwrapping it.

```suggestion
	oneTimeCallbackRef.current = oneTimeCallback;
```

File: client/sites-dashboard/components/sites-table-row.tsx (hunk 3 above)
Comment (zaguiini) [inline client/sites-dashboard/components/sites-table-row.tsx]: We can revert this change 😁

File: client/lib/use-in-view/index.ts [hunk 6]
```diff
@@ -27,20 +26,21 @@ export function useInView< T extends Element >( oneTimeCallback: () => void ): R
 				return;
 			}
 
-			oneTimeCallback();
-
+			oneTimeCallbackRef.current?.();
 			viewedRef.current = true;
+			// eslint-disable-next-line @typescript-eslint/no-use-before-define
+			observer.disconnect();
```
Comment (zaguiini) [inline client/lib/use-in-view/index.ts]: We can remove this linter error if we inline the `handler` function as an argument to the constructor instead of defining and passing the reference.

File: client/lib/use-in-view/index.ts [hunk 7]
```diff
@@ -1,21 +1,20 @@
...
 		}
 
-		// If the observer is already defined, no need to continue
-		if ( observerRef.current ) {
-			return;
-		}
-
 		const handler = ( entries: IntersectionObserverEntry[] ) => {
 			// Only fire once per page load
 			if ( viewedRef.current ) {
```
Comment (zaguiini) [inline client/lib/use-in-view/index.ts]: Can we get rid of this `viewedRef`? We want to fire the callback only once and it's triggered by the intersection observer, so isn't it a matter of disconnecting the observer once the callback is called, as we're doing that already by calling `observer.disconnect()`?

In fact, we don't need to declare the refs as dependencies of the `useEffect` array since the ref wrappers itself are stable -- the linter doesn't complain 😄 So we won't see the effect running multiple times and disconnecting it will prevent it from running several times.

**Review by zaguiini — COMMENTED** [review]
We're almost there!

Question: are we having both our internal solution `useInView` and the dependency at the same time? They're named the same, but ours only run once while theirs run indefinitely unless explicitly defined by the `triggerOnce` option. Are we streamlining that anytime soon?

File: client/lib/use-in-view/index.ts (hunk 5 above)
Comment (zaguiini) [reply client/lib/use-in-view/index.ts]: Fixed.

File: client/sites-dashboard/components/sites-table-row.tsx (hunk 3 above)
Comment (zaguiini) [reply client/sites-dashboard/components/sites-table-row.tsx]: Fixed.

File: client/lib/use-in-view/index.ts (hunk 6 above)
Comment (zaguiini) [reply client/lib/use-in-view/index.ts]: Fixed.

File: client/lib/use-in-view/index.ts (hunk 7 above)
Comment (zaguiini) [reply client/lib/use-in-view/index.ts]: Fixed.

**Comment by zaguiini** [comment]
@Automattic/build I have applied changes since @danielbachhuber wants to land this ASAP.

One thing that's left is that we still have usages of `react-intersection-observer` in our codebase. I believe we want to get rid of the dependency entirely, right? So are we having a follow-up that replaces the remaining usages? Did we achieve feature parity already?

File: client/lib/use-in-view/test/index.tsx (hunk 4 above)
Comment (danielbachhuber) [reply client/lib/use-in-view/test/index.tsx:85]: 👍

**Review by danielbachhuber — APPROVED** [review]
Works great, thanks!

Landing so we can fix the `calypso_sites_dashboard_plan_renew_nag_inview` Tracks event.

**Comment by p-jackson** [comment]
>  So are we having a follow-up that replaces the remaining usages? Did we achieve feature parity already?

This should remove the last use in Calypso #69469
